include/crocoddyl/multibody/actions/contact-invdyn.hpp

```diff
@@ -0,0 +1,618 @@
+
+namespace crocoddyl {
+
+/**
+ * @brief Differential action model for contact inverse dynamics in multibody systems.
+ *
+ * This class implements forward kinematic with contact holonomic constraints (defined at the acceleration level) and
+ * inverse-dynamics computation using the Recursive Newton Euler Algorithm (RNEA). The stack of cost and constraint
+ * functions are implemented in `CostModelSumTpl` and `ConstraintModelManagerTpl`, respectively.
+ * The acceleration and contact forces are decision variables defined as the control inputs, and the under-actuation
+ * and contact constraint are under the name `tau` and its frame name, thus the user is not allow to use it.
+ *
+ * Additionally, it is important remark that `calcDiff()` computes the derivatives using the latest stored values by
+ * `calc()`. Thus, we need to run first `calc()`.
+ *
+ * \sa `DifferentialActionModelAbstractTpl`, `calc()`, `calcDiff()`, `createData()`
+ */
+template <typename _Scalar>
+class DifferentialActionModelContactInvDynamicsTpl : public DifferentialActionModelAbstractTpl<_Scalar> {
+ public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef DifferentialActionModelAbstractTpl<Scalar> Base;
+  typedef DifferentialActionDataContactInvDynamicsTpl<Scalar> Data;
+  typedef StateMultibodyTpl<Scalar> StateMultibody;
+  typedef ActuationModelAbstractTpl<Scalar> ActuationModelAbstract;
+  typedef CostModelSumTpl<Scalar> CostModelSum;
+  typedef ConstraintModelManagerTpl<Scalar> ConstraintModelManager;
+  typedef ContactModelMultipleTpl<Scalar> ContactModelMultiple;
+  typedef DataCollectorAbstractTpl<Scalar> DataCollectorAbstract;
+  typedef DifferentialActionDataAbstractTpl<Scalar> DifferentialActionDataAbstract;
+  typedef ContactItemTpl<Scalar> ContactItem;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  /**
+   * @brief Initialize the contact inverse-dynamics action model
+   *
+   * It describes the kinematic evolution of the multibody system with contacts,
+   * and computes the needed torques using inverse-dynamics.
+   *
+   * @param[in] state      State of the multibody system
+   * @param[in] actuation  Actuation model
+   * @param[in] contacts   Multiple contacts
+   * @param[in] costs      Cost model
+   */
+  DifferentialActionModelContactInvDynamicsTpl(boost::shared_ptr<StateMultibody> state,
+                                               boost::shared_ptr<ActuationModelAbstract> actuation,
+                                               boost::shared_ptr<ContactModelMultiple> contacts,
+                                               boost::shared_ptr<CostModelSum> costs);
+
+  /**
+   * @brief Initialize the contact inverse-dynamics action model
+   *
+   * @param[in] state        State of the multibody system
+   * @param[in] actuation    Actuation model
+   * @param[in] contacts   Multiple contacts
```

Comment:
Formatting. I wonder why it passed CI

Reply:
This is not tested by clang-format. However, I have fixed this thing.